### PR TITLE
Use runnableExamples for nre example code where possible

### PR DESCRIPTION
**Pull request by flaviut**
Thanks to @timotheecour [for the suggestion!](https://github.com/nim-lang/Nim/commit/a0ac0b9696f3a79fbde153c63878aea363b09ad4#commitcomment-30755959)

This PR is blocked on  #9170

### CODE REVIEW

**Comment by flaviut** [comment]
This PR has been updated to not use runnableExamples in a type docstring. It would still be nice if #9170 was implemented, so that it is possible to use runnableExamples there.

File: lib/impure/nre.nim [hunk 1]
```diff
@@ -495,8 +492,12 @@ proc matchImpl(str: string, pattern: Regex, start, endpos: int, flags: int): Opt
 
 proc match*(str: string, pattern: Regex, start = 0, endpos = int.high): Option[RegexMatch] =
   ## Like ```find(...)`` <#proc-find>`_, but anchored to the start of the
-  ## string. This means that ``"foo".match(re"f") == true``, but
-  ## ``"foo".match(re"o") == false``.
+  ## string.
+  ##
```
Comment (timotheecour) [inline lib/impure/nre.nim]: nit: is this empty `##` line needed? (+ elsewhere)

File: lib/impure/nre.nim [hunk 2]
```diff
@@ -40,29 +40,26 @@ export options
...
+    1 .. 1,
+    2 .. 2,
+    4 .. 4,
+    6 .. 6,
+    7 .. 7,
+  ]
+  var i = 0
+  for match in "moigagoo".findIter(vowels):
+    doAssert match.matchBounds == expectedResults[i]
+    i += 1
```
Comment (timotheecour) [inline lib/impure/nre.nim]: i.inc?

**Comment by timotheecour** [comment]
> This PR has been updated to not use runnableExamples in a type docstring. It would still be nice if #9170 was implemented, so that it is possible to use runnableExamples there.

as a workaround until #9170  is implemented, how about adding `runnableExamples` right below the type definition?
(it'd become a top-level `runnableExamples`)

File: lib/impure/nre.nim (hunk 1 above)
Comment (flaviut) [reply lib/impure/nre.nim]: Probably not, but I like the way it looks and separates the example from the rest of the docstring.

File: lib/impure/nre.nim (hunk 2 above)
Comment (flaviut) [reply lib/impure/nre.nim]: done

**Comment by flaviut** [comment]
>as a workaround until #9170 is implemented, how about adding runnableExamples right below the type definition?

That's a good idea. I was reluctant to do that due to the code duplication, but I've changed my mind after thinking about it more.

**Comment by timotheecour** [comment]
> That's a good idea. I was reluctant to do that due to the code duplication, but I've changed my mind after thinking about it more.

I meant instead rather than in addition to; ya, code duplication is bad.

**Comment by dom96** [comment]
Looks good. Note: we will likely remove ``nre`` from stdlib (https://github.com/nim-lang/Nim/issues/7278)

**Pull request merged**